query.go

```
@@ -0,0 +1,53 @@
+package ep
+
+import (
+	"context"
+	"github.com/satori/go.uuid"
+	"time"
+)
+
+var _ = registerGob(&Query{}, QueryType(0))
+
+// QueryType uint to specify all types
+type QueryType uint
+
+// Query stores query metadata
+type Query struct {
+	ID             string
+	Typee          QueryType
+	ctx            context.Context
+	StartedAt      time.Time
+	MasterAddr     string
+	signalChannels []chan error
+}
+
+// SetDynamicFields set the static fields of the current Query
+func (q *Query) SetDynamicFields(ctx context.Context) {
+	q.ctx = ctx
+}
+
+// CopyStaticQueryFields return a new Query while initialize only the static fields of the current Query
+func (q *Query) CopyStaticQueryFields() *Query {
+	return &Query{
+		ID:         q.ID,
+		Typee:      q.Typee,
+		MasterAddr: q.MasterAddr,
+		StartedAt:  q.StartedAt,
+	}
+}
+
+// NewQuery return a new Query object from a given args
+func NewQuery(ctx context.Context, masterAddr string) (*Query, error) {
+	uid, err := uuid.NewV4()
+	queryType, ok := ctx.Value(QueryTypeKey).(QueryType)
+	if !ok {
+		queryType = 0
+	}
+	return &Query{
+		ID:         uid.String(),
+		Typee:      queryType,
+		ctx:        ctx,
+		MasterAddr: masterAddr,
+		StartedAt:  time.Now(),
+	}, err
```

Why do we return an error here? If it happens, it happens long before `return` statement. We also use the positive result of `uuid.NewV4` ignoring the error, might it be a problem?